### DISCUSSION TITLE
Struggling with construct_params in compiler.py

### DISCUSSION THREAD

**Discussion by andrewcooke** [sqlalchemy/sqlalchemy#6377] [Ideas]
This would be an issue, but I don't yet have a nice example so I am starting a discussion in case this turns out to be obvious or a mistake on my part (the query is so complex that constructing a nice example is going to take some time).  I'm a bit out of my depth, so what follows may contain some mistakes, sorry.

The bug occurs when the (complex) statement re-uses the same bind parameter twice (a repeated subquery).  The first time the query is executed, everything is fine.  But on subsequent calls the `orig_extracted` parameter seems to be used in one place (and the correct `extracted_parameters` in the other).  So if subsequent calls use a different parameter the query sent to the database is incorrect (one of the parameter uses is correct, but the other is from the original call when the query was first compiled).

Does that make any sense at all?  I can see this happening in the debugger, and the database logger is clearly showing the wrong parameter, but I don't really understand the code.

My problem with creating an example is that the SQL query is constructed automatically given a parse tree for a separate query language.  Reducing that logic is going to take some time.

The query is:

```
WITH anon_1 AS 
(SELECT anon_2.activity_journal_id AS activity_journal_id 
FROM (SELECT activity_journal.id AS activity_journal_id 
FROM source JOIN activity_journal ON source.id = activity_journal.id 
WHERE activity_journal.id IN (SELECT anon_3.id 
FROM (SELECT source.id AS id 
FROM source JOIN (statistic_journal JOIN statistic_journal_timestamp ON statistic_journal.id = statistic_journal_timestamp.id) ON source.id = statistic_journal.source_id JOIN statistic_name ON statistic_name.id = statistic_journal.statistic_name_id 
WHERE statistic_name.name LIKE %(name_1)s AND statistic_journal_timestamp.value > %(value_1)s) AS anon_3) UNION SELECT activity_journal.id AS activity_journal_id 
FROM source JOIN activity_journal ON source.id = activity_journal.id JOIN file_hash ON file_hash.id = activity_journal.file_hash_id JOIN (source AS source_1 JOIN activity_topic_journal AS activity_topic_journal_1 ON source_1.id = activity_topic_journal_1.id) ON file_hash.id = activity_topic_journal_1.file_hash_id 
WHERE activity_topic_journal_1.id IN (SELECT anon_4.id 
FROM (SELECT source.id AS id 
FROM source JOIN (statistic_journal JOIN statistic_journal_timestamp ON statistic_journal.id = statistic_journal_timestamp.id) ON source.id = statistic_journal.source_id JOIN statistic_name ON statistic_name.id = statistic_journal.statistic_name_id 
WHERE statistic_name.name LIKE %(name_2)s AND statistic_journal_timestamp.value > %(value_2)s) AS anon_4)) AS anon_2)
 SELECT source.id AS source_id, source.type AS source_type, source.activity_group_id AS source_activity_group_id 
FROM source 
WHERE source.id IN (SELECT anon_1.activity_journal_id 
FROM anon_1)
```

where name_1/value_1 and name_2/value_2 are from the same query and should take the same value, but name_2 (i think) is being given the value from the initial compilation.

**Comment by zzzeek** [comment]
the problem goes away if you set [query_cache_size=0](https://docs.sqlalchemy.org/en/14/core/engines.html?highlight=query_cache_size#sqlalchemy.create_engine.params.query_cache_size)  ?  Generally if you are working with `@compiles`, caching should not be used by default because to be compatible w caching requires extra directives that aren't really documented right now.  There's a flag that shuts it off for that construct.   So...that would be not working here, what does your custom class look like?

**Reply by zzzeek** [reply]
I can think of some ways this might be working right now, which I should get a fix out ASAP but show me what you are subclassing.

**Reply by andrewcooke** [reply]
i am not subclassing anything.  i am building the SQL that i edited in above and then calling query.all().

the only reason i am looking at the compiler code is in the debugger to work out what is happening.  i am not modifying the compiler.  sorry, that probably wasn't clear.

**Reply by andrewcooke** [reply]
trying the cache size thing now

**Reply by andrewcooke** [reply]
YES!  that fixes it.

**Reply by zzzeek** [reply]
ok that is **very bad**, please provide an [MCVE](http://stackoverflow.com/help/mcve) as soon as possible, caching failing is an **extremely critical** issue and I will literally work through the night to release a fix, thanks for assisting

**Reply by zzzeek** [reply]
are you using bindparam() ?  can you please share the code used to produce the query ?    ASAP please??

**Reply by zzzeek** [reply]
SQLAlchemy 1.4.11 ????    or are you on an old 1.4 version???  there were some issues fixed early on

**Reply by andrewcooke** [reply]
ok, i'll make an example.  i am not using bindparam.  i think the problem may be related to re-using a query.  i'll try get something working tonight.

**Reply by andrewcooke** [reply]
1.4.11

**Reply by andrewcooke** [reply]
Is there some way in SQLAlchemy to see the query passed to the database?  What has been causing me problems is that the params actually passed are different from `q.statement.compile(dialect=postgresql.dialect()).params`.  So it is hard to be sure the bug occurs without looking at the satabase logs.

**Reply by CaselIT** [reply]
add `echo=True` in create engine

**Reply by zzzeek** [reply]
name_1 and name_2 are different, but are from the "same" query, I wonder how you'd do that?

**Reply by andrewcooke** [reply]
so do i.  it's not easy to duplicate.  bear with me...

**Reply by andrewcooke** [reply]
ok, this is a huge amount of work.  i have tried to reproduce it using a simpler set of tables and failed.  so i need to take my existing code and reduce it.  it will probably be days.  so don't wait up.

**Reply by zzzeek** [reply]
Ok, can I see the simple version?  I can look to see what might cause a problem under certain conditions.  Repro for cache is very difficult so I just need to get the general idea.

**Reply by andrewcooke** [reply]
i deleted it, but it was roughly

```
with Session() as s:
    q1 = s.query(Table.id).filter(Table.id == N)
    q2 = s.query(Table.id).filter(Table.id.in_(q1.subquery())
    q3 = s.query(Table.id).filter(Table.id.in_(q1.subquery())
    q4 = s.query(Table).filter(Table.id.in(union(q2, q3)))
    q4.all()
```

run with N=1 and then N=2.  in the second run a value of 1 used for one parameter.  except that it wasn't ( icouldn;t reproduce the prblem).  the actual code uses a similar structure but with table inheritance and more joins.  so i am trying to pull out all the table definitions i am using.

**Comment by andrewcooke** [comment]
there's a first test case.  I WILL SIMPLIFY THIS.  but it shows what is happening and i think you can run it stand-alone already if you cut _ paste to somewhere else.   https://github.com/andrewcooke/choochoo/blob/786a038eb0aa4cc236a7cfa1b628adaa636509a5/py/tests/test_search.py#L71 the output from postgresql (new instance running in docker with log_statement=all is:

```
2021-04-28 13:54:49.217 GMT [92] LOG:  statement: WITH anon_1 AS 
        (SELECT anon_2.activity_journal_id AS activity_journal_id 
        FROM (SELECT activity_journal.id AS activity_journal_id 
        FROM source JOIN activity_journal ON source.id = activity_journal.id 
        WHERE activity_journal.id IN (SELECT anon_3.id 
        FROM (SELECT source.id AS id 
        FROM source JOIN (statistic_journal JOIN statistic_journal_timestamp ON statistic_journal.id = statistic_journal_timestamp.id) ON source.id = statistic_journal.source_id JOIN statistic_name ON statistic_name.id = statistic_journal.statistic_name_id 
        WHERE statistic_name.name LIKE 'start' AND statistic_journal_timestamp.value > '2020-01-01T03:00:00+00:00'::timestamptz) AS anon_3) UNION SELECT activity_journal.id AS activity_journal_id 
        FROM source JOIN activity_journal ON source.id = activity_journal.id JOIN file_hash ON file_hash.id = activity_journal.file_hash_id JOIN (source AS source_1 JOIN activity_topic_journal AS activity_topic_journal_1 ON source_1.id = activity_topic_journal_1.id) ON file_hash.id = activity_topic_journal_1.file_hash_id 
        WHERE activity_topic_journal_1.id IN (SELECT anon_4.id 
        FROM (SELECT source.id AS id 
        FROM source JOIN (statistic_journal JOIN statistic_journal_timestamp ON statistic_journal.id = statistic_journal_timestamp.id) ON source.id = statistic_journal.source_id JOIN statistic_name ON statistic_name.id = statistic_journal.statistic_name_id 
        WHERE statistic_name.name LIKE 'start' AND statistic_journal_timestamp.value > '2020-01-01T03:00:00+00:00'::timestamptz) AS anon_4)) AS anon_2)
         SELECT source.id AS source_id, source.type AS source_type 
        FROM source 
        WHERE source.id IN (SELECT anon_1.activity_journal_id 
        FROM anon_1)
2021-04-28 13:54:49.225 GMT [92] LOG:  statement: WITH anon_1 AS 
        (SELECT anon_2.activity_journal_id AS activity_journal_id 
        FROM (SELECT activity_journal.id AS activity_journal_id 
        FROM source JOIN activity_journal ON source.id = activity_journal.id 
        WHERE activity_journal.id IN (SELECT anon_3.id 
        FROM (SELECT source.id AS id 
        FROM source JOIN (statistic_journal JOIN statistic_journal_timestamp ON statistic_journal.id = statistic_journal_timestamp.id) ON source.id = statistic_journal.source_id JOIN statistic_name ON statistic_name.id = statistic_journal.statistic_name_id 
        WHERE statistic_name.name LIKE 'start' AND statistic_journal_timestamp.value > '2020-01-01T03:00:00+00:00'::timestamptz) AS anon_3) UNION SELECT activity_journal.id AS activity_journal_id 
        FROM source JOIN activity_journal ON source.id = activity_journal.id JOIN file_hash ON file_hash.id = activity_journal.file_hash_id JOIN (source AS source_1 JOIN activity_topic_journal AS activity_topic_journal_1 ON source_1.id = activity_topic_journal_1.id) ON file_hash.id = activity_topic_journal_1.file_hash_id 
        WHERE activity_topic_journal_1.id IN (SELECT anon_4.id 
        FROM (SELECT source.id AS id 
        FROM source JOIN (statistic_journal JOIN statistic_journal_timestamp ON statistic_journal.id = statistic_journal_timestamp.id) ON source.id = statistic_journal.source_id JOIN statistic_name ON statistic_name.id = statistic_journal.statistic_name_id 
        WHERE statistic_name.name LIKE 'start' AND statistic_journal_timestamp.value > '2021-01-01T03:00:00+00:00'::timestamptz) AS anon_4)) AS anon_2)
         SELECT source.id AS source_id, source.type AS source_type 
        FROM source 
        WHERE source.id IN (SELECT anon_1.activity_journal_id 
        FROM anon_1)
```

and you can see that the second query has both 2020 and 2021.

again, i will simplify this, but that is a working standalone example.

**Reply by andrewcooke** [reply]
probably tricial but i use https://github.com/andrewcooke/choochoo/blob/dev/dkr/run-pg-transient.sh with `--log` to start a docker instance.

**Reply by andrewcooke** [reply]
Simplified query at https://github.com/andrewcooke/choochoo/blob/ecedd05f13c59b80ae779bb80f797ba7b0d557f2/py/tests/test_search.py#L71

```
            q = s.query(Source.id). \
                join(StatisticJournalTimestamp). \
                join(StatisticName). \
                filter(StatisticName.name.like('start')). \
                filter(StatisticJournalTimestamp.value > value)
            q_direct = s.query(ActivityJournal.id). \
                filter(ActivityJournal.id.in_(q.subquery()))
            q_via_topic = s.query(ActivityJournal.id). \
                join(FileHash). \
                join(ActivityTopicJournal). \
                filter(ActivityTopicJournal.id.in_(q.subquery()))
            constraints = union(q_direct, q_via_topic).subquery().select()
            return s.query(Source).filter(Source.id.in_(constraints))
```

**Reply by andrewcooke** [reply]
STANDALONE TEST IN SINGLE FILE - https://github.com/andrewcooke/choochoo/blob/dev/py/tests/test_sqlalchemy.py

**Reply by andrewcooke** [reply]
Note the test does not fail.  I still have no way to PROGRAMMATICALLY get the output sent to the database (echo suggestion above doesn't help me write a test).  You need to look at the queries logged to postgres.

**Reply by zzzeek** [reply]
great, I've extracted this into standalone and it reproduces.  will open the issue now this will be released today , thanks for all the effort to create this test case!!!

**Reply by andrewcooke** [reply]
brilliant.  thanks for waiting.  and thanks for the software - it's by far the nicest db library i have used,

**Reply by zzzeek** [reply]
thanks for your help.  the issue is #6391 and it's merged to master.  Can you try out git master if you have a chance (and turn the cache back on?)  thanks.

**Reply by andrewcooke** [reply]
i tried, but if i change 'sqlalchemy' for 'sqlalchemy @ git+https://github.com/sqlalchemy/sqlalchemy@master#egg=sqlalchemy' in setup.py pip throws a fit on conflicting requirements for, i think, colorlog.  trying to understand what is happening.

**Reply by andrewcooke** [reply]
had to downgrade pip; now have it installed.  this issue is resolved (parameters are now correct).  thanks,

**Reply by CaselIT** [reply]
> 
> 
> i tried, but if i change 'sqlalchemy' for 'sqlalchemy @ git+https://github.com/sqlalchemy/sqlalchemy@master#egg=sqlalchemy' in setup.py pip throws a fit on conflicting requirements for, i think, colorlog. trying to understand what is happening.

sadly that's pip doing it's usual thing. We also have random conflicts. I think it's because somewhere it cannot resolve the dev versions properly so in our case complains that sqlalchemy 1.4.x.dev < 1.3.x or some similar error

This is the relevant pypi issue https://github.com/pypa/pip/issues/9446

**Reply by andrewcooke** [reply]
Yes, it was conflicting with GeoSQLAlchemy's requirements because of the dev.
